File: src/test/java/org/stt/model/TimeTrackingItemTest.java
package org.stt.model;

import org.junit.Test;
import org.stt.time.DateTimes;

import java.time.LocalDateTime;
import java.util.Optional;

import static java.time.LocalDateTime.now;
import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.assertThat;

public class TimeTrackingItemTest {

	@Test
	public void shouldCreateItemWithoutEnd() {
		// GIVEN

		// WHEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now());

		// THEN
		assertThat(sut.getEnd().isPresent(), is(false));
	}

	@Test
	public void shouldCreateItemWithStartIfEndIsMissing() {
		// GIVEN

		// WHEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now());

		// THEN
		assertThat(sut.getStart(), notNullValue());
	}

	@Test
	public void shouldCreateItemWithStart() {
		// GIVEN

		// WHEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now(),
                LocalDateTime.now().plusMinutes(1));

		// THEN
		assertThat(sut.getStart(), notNullValue());
	}

	@Test
	public void withEndShouldCreateNewItem() {
		// GIVEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now());

		// WHEN
        LocalDateTime newEndTime = DateTimes.preciseToSecond(now().plusMinutes(2));
        TimeTrackingItem newItem = sut.withEnd(newEndTime);

		// THEN
		assertThat(newItem, not(is(sut)));
		assertThat(sut.getEnd().isPresent(), is(false));
		assertThat(newItem.getEnd().get(), is(newEndTime));
	}

	@Test
    public void withStartShouldCreateNewItem() {
        // GIVEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now());

        // WHEN
        LocalDateTime newStartTime = DateTimes.preciseToSecond(now().plusMinutes(2));
        TimeTrackingItem newItem = sut.withStart(newStartTime);

        // THEN
        assertThat(newItem, not(is(sut)));
        assertThat(newItem.getStart(), is(newStartTime));
    }

    @Test
    public void withPendingEndShouldCreateNewItem() {
        // GIVEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        TimeTrackingItem newItem = sut.withPendingEnd();

        // THEN
        assertThat(newItem, not(is(sut)));
        assertThat(newItem.getEnd(), is(Optional.empty()));
    }

    @Test
    public void withActivityShouldCreateNewItem() {
        // GIVEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        TimeTrackingItem newItem = sut.withActivity("11");

        // THEN
        assertThat(newItem, not(is(sut)));
        assertThat(newItem.getActivity(), is("11"));
    }

    @Test
    public void withActivityShouldCreateNewItemForOngoing() {
        // GIVEN
        TimeTrackingItem sut = new TimeTrackingItem("", LocalDateTime.now());

        // WHEN
        TimeTrackingItem newItem = sut.withActivity("11");

        // THEN
        assertThat(newItem, not(is(sut)));
        assertThat(newItem.getActivity(), is("11"));
    }

    @Test
    public void shouldReturnTrueForSameStart() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = a.withEnd(LocalDateTime.now().plusDays(2));

        // WHEN
        boolean result = a.sameStartAs(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseForDifferentStart() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1));

        // WHEN
        boolean result = a.sameStartAs(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnTrueForSameActivity() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("Activity", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1));

        // WHEN
        boolean result = a.sameActivityAs(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseForDifferentActivity() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("Activity", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1)).withActivity("Other");

        // WHEN
        boolean result = a.sameActivityAs(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnTrueForSameEnd() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1));

        // WHEN
        boolean result = a.sameEndAs(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueForBothOngoing() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now());
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1));

        // WHEN
        boolean result = a.sameEndAs(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseForOnlyOneOngoing() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now());
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1)).withEnd(LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.sameEndAs(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnFalseForDifferentEnds() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now()).withEnd(LocalDateTime.now().plusDays(2));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusSeconds(1)).withEnd(LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.sameEndAs(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnTrueForPartialOverlap() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now()).withEnd(LocalDateTime.now().plusDays(2));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusDays(1)).withEnd(LocalDateTime.now().plusDays(2));

        // WHEN
        boolean result = a.intersects(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueForEnclosingInterval() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now()).withEnd(LocalDateTime.now().plusDays(3));
        TimeTrackingItem b = a.withStart(LocalDateTime.now().plusDays(1)).withEnd(LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.intersects(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueForEqualInterval() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now()).withEnd(LocalDateTime.now().plusDays(3));

        // WHEN
        boolean result = a.intersects(a);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseForStartMatchingOtherEnd() {
        // GIVEN
        LocalDateTime end = LocalDateTime.now().plusDays(1);
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now()).withEnd(end);
        TimeTrackingItem b = a.withEnd(LocalDateTime.now().plusDays(2)).withStart(end);

        // WHEN
        boolean result = a.intersects(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnTrueIfOngoingAndOtherIsNot() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now());
        TimeTrackingItem b = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.endsSameOrAfter(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueIfBothAreOngoing() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now());
        TimeTrackingItem b = new TimeTrackingItem("", LocalDateTime.now());

        // WHEN
        boolean result = a.endsSameOrAfter(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueIfEndsAfterOther() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(2));
        TimeTrackingItem b = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.endsSameOrAfter(b);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseIfEndsBeforeOther() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));
        TimeTrackingItem b = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(2));

        // WHEN
        boolean result = a.endsSameOrAfter(b);

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnTrueIfEndsAtSameTime() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.endsSameOrAfter(a);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueIfEndsBeforeDate() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(1));

        // WHEN
        boolean result = a.endsAtOrBefore(LocalDateTime.now().plusDays(2));

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnTrueIfEndsAtDate() {
        // GIVEN
        LocalDateTime end = LocalDateTime.now().plusDays(1);
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), end);

        // WHEN
        boolean result = a.endsAtOrBefore(end);

        // THEN
        assertThat(result, is(true));
    }

    @Test
    public void shouldReturnFalseIfEndsAfter() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now(), LocalDateTime.now().plusDays(2));

        // WHEN
        boolean result = a.endsAtOrBefore(LocalDateTime.now().plusDays(1));

        // THEN
        assertThat(result, is(false));
    }

    @Test
    public void shouldReturnFalseIfOngoing() {
        // GIVEN
        TimeTrackingItem a = new TimeTrackingItem("", LocalDateTime.now());

        // WHEN
        boolean result = a.endsAtOrBefore(LocalDateTime.now());

        // THEN
        assertThat(result, is(false));
    }

}

File: src/test/kotlin/org/stt/model/TimeTrackingItemTest.kt
package org.stt.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.stt.time.preciseToSecond
import java.time.LocalDateTime
import java.time.LocalDateTime.now

class TimeTrackingItemTest {

    @Test
    fun shouldCreateItemWithoutEnd() {
        // GIVEN

        // WHEN
        val (_, _, end) = TimeTrackingItem("", now())

        // THEN
        assertThat(end).isNull()
    }

    @Test
    fun shouldCreateItemWithStartIfEndIsMissing() {
        // GIVEN

        // WHEN
        val (_, start) = TimeTrackingItem("", now())

        // THEN
        assertThat(start).isNotNull()
    }

    @Test
    fun shouldCreateItemWithStart() {
        // GIVEN

        // WHEN
        val (_, start) = TimeTrackingItem("", now(),
                now().plusMinutes(1))

        // THEN
        assertThat(start).isNotNull()
    }

    @Test
    fun withEndShouldCreateNewItem() {
        // GIVEN
        val sut = TimeTrackingItem("", now())

        // WHEN
        val newEndTime = now().preciseToSecond().plusMinutes(2)
        val newItem = sut.withEnd(newEndTime)

        // THEN
        assertThat(newItem).isNotEqualTo(sut)
        assertThat(sut.end).isNull()
        assertThat(newItem.end).isEqualTo(newEndTime)
    }

    @Test
    fun withStartShouldCreateNewItem() {
        // GIVEN
        val sut = TimeTrackingItem("", now())

        // WHEN
        val newStartTime = now().plusMinutes(2).preciseToSecond()
        val newItem = sut.withStart(newStartTime)

        // THEN
        assertThat(newItem).isNotEqualTo(sut)
        assertThat(newItem.start).isEqualTo(newStartTime)
    }

    @Test
    fun withPendingEndShouldCreateNewItem() {
        // GIVEN
        val sut = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val newItem = sut.withPendingEnd()

        // THEN
        assertThat(newItem).isNotEqualTo(sut)
        assertThat(newItem.end).isNull()
    }

    @Test
    fun withActivityShouldCreateNewItem() {
        // GIVEN
        val sut = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val newItem = sut.withActivity("11")

        // THEN
        assertThat(newItem).isNotEqualTo(sut)
        assertThat(newItem.activity).isEqualTo("11")
    }

    @Test
    fun withActivityShouldCreateNewItemForOngoing() {
        // GIVEN
        val sut = TimeTrackingItem("", now())

        // WHEN
        val newItem = sut.withActivity("11")

        // THEN
        assertThat(newItem).isNotEqualTo(sut)
        assertThat(newItem.activity).isEqualTo("11")
    }

    @Test
    fun shouldReturnTrueForSameStart() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))
        val b = a.withEnd(now().plusDays(2))

        // WHEN
        val result = a.sameStartAs(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseForDifferentStart() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))
        val b = a.withStart(now().plusSeconds(1))

        // WHEN
        val result = a.sameStartAs(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnTrueForSameActivity() {
        // GIVEN
        val a = TimeTrackingItem("Activity", now(), now().plusDays(1))
        val b = a.withStart(now().plusSeconds(1))

        // WHEN
        val result = a.sameActivityAs(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseForDifferentActivity() {
        // GIVEN
        val a = TimeTrackingItem("Activity", now(), now().plusDays(1))
        val b = a.withStart(now().plusSeconds(1)).withActivity("Other")

        // WHEN
        val result = a.sameActivityAs(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnTrueForSameEnd() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))
        val b = a.withStart(now().plusSeconds(1))

        // WHEN
        val result = a.sameEndAs(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueForBothOngoing() {
        // GIVEN
        val a = TimeTrackingItem("", now())
        val b = a.withStart(now().plusSeconds(1))

        // WHEN
        val result = a.sameEndAs(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseForOnlyOneOngoing() {
        // GIVEN
        val a = TimeTrackingItem("", now())
        val b = a.withStart(now().plusSeconds(1)).withEnd(now().plusDays(1))

        // WHEN
        val result = a.sameEndAs(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnFalseForDifferentEnds() {
        // GIVEN
        val a = TimeTrackingItem("", now()).withEnd(now().plusDays(2))
        val b = a.withStart(now().plusSeconds(1)).withEnd(now().plusDays(1))

        // WHEN
        val result = a.sameEndAs(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnTrueForPartialOverlap() {
        // GIVEN
        val a = TimeTrackingItem("", now()).withEnd(now().plusDays(2))
        val b = a.withStart(now().plusDays(1)).withEnd(now().plusDays(2))

        // WHEN
        val result = a.intersects(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueForEnclosingInterval() {
        // GIVEN
        val a = TimeTrackingItem("", now()).withEnd(now().plusDays(3))
        val b = a.withStart(now().plusDays(1)).withEnd(now().plusDays(1))

        // WHEN
        val result = a.intersects(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueForEqualInterval() {
        // GIVEN
        val a = TimeTrackingItem("", now()).withEnd(now().plusDays(3))

        // WHEN
        val result = a.intersects(a)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseForStartMatchingOtherEnd() {
        // GIVEN
        val end = now().plusDays(1)
        val a = TimeTrackingItem("", now()).withEnd(end)
        val b = a.withEnd(now().plusDays(2)).withStart(end)

        // WHEN
        val result = a.intersects(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnTrueIfOngoingAndOtherIsNot() {
        // GIVEN
        val a = TimeTrackingItem("", now())
        val b = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val result = a.endsSameOrAfter(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueIfBothAreOngoing() {
        // GIVEN
        val a = TimeTrackingItem("", now())
        val b = TimeTrackingItem("", now())

        // WHEN
        val result = a.endsSameOrAfter(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueIfEndsAfterOther() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(2))
        val b = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val result = a.endsSameOrAfter(b)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseIfEndsBeforeOther() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))
        val b = TimeTrackingItem("", now(), now().plusDays(2))

        // WHEN
        val result = a.endsSameOrAfter(b)

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnTrueIfEndsAtSameTime() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val result = a.endsSameOrAfter(a)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueIfEndsBeforeDate() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(1))

        // WHEN
        val result = a.endsAtOrBefore(now().plusDays(2))

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnTrueIfEndsAtDate() {
        // GIVEN
        val end = now().plusDays(1)
        val a = TimeTrackingItem("", now(), end)

        // WHEN
        val result = a.endsAtOrBefore(end)

        // THEN
        assertThat(result).isTrue()
    }

    @Test
    fun shouldReturnFalseIfEndsAfter() {
        // GIVEN
        val a = TimeTrackingItem("", now(), now().plusDays(2))

        // WHEN
        val result = a.endsAtOrBefore(now().plusDays(1))

        // THEN
        assertThat(result).isFalse()
    }

    @Test
    fun shouldReturnFalseIfOngoing() {
        // GIVEN
        val a = TimeTrackingItem("", now())

        // WHEN
        val result = a.endsAtOrBefore(now())

        // THEN
        assertThat(result).isFalse()
    }

}
